panel/applets/icon-tasklist/WindowListView.vala

```diff
@@ -0,0 +1,149 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright (C) 2016 Fernando Mussel <fernandomussel91@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+public class WindowListView
+{
+
+    private unowned WindowManager? window_manager = null;
+
+    private Gtk.Popover? window_list_popover = null;
+    private Gtk.ListBox window_listbox;
+    private unowned Gtk.Widget owner;
+    private HashTable<unowned ApplicationWindow, unowned WindowListItem>? window_row_map = null;
+    private Gtk.SizeGroup switch_btn_group;
+    private Queue<Gtk.ListBoxRow> removed_items;
+
+    public WindowListView(Gtk.Widget owner, WindowManager window_manager)
+    {
+        this.owner = owner;
+        this.window_manager = window_manager;
+        window_row_map = new HashTable<unowned ApplicationWindow,unowned WindowListItem>(direct_hash, direct_equal);
+        configure_view();
+        removed_items = new Queue<Gtk.ListBoxRow>();
+    }
+
+    ~WindowListView()
+    {
+        clear_removed_items();
+    }
+
+    public uint get_window_num()
+    {
+        return window_row_map.size();
+    }
+
+    /**
+     * Add new WindowListItem associated with \p appwin
+     */
+    public void add_window(ApplicationWindow appwin)
+    {
+        WindowListItem win_row = add_window_list_item(appwin);
+
+        window_row_map [appwin] = win_row;
+        // set the row object as the listener for the window's name change
+        // event
+        appwin.set_name_change_listener(win_row);
+    }
+
+    /**
+     * Handles the view side operations required to add a new ListBoxRow associated
+     * with the supplied window
+     */
+    private WindowListItem add_window_list_item(ApplicationWindow appwin)
+    {
+        WindowListItem new_item = new WindowListItem(appwin.get_window().get_name(),
+                                                     appwin, window_manager);
+
+        // this ensures that the switch buttons of all WindowListItems will have
+        // the same length
+        new_item.add_switch_button_to_group(switch_btn_group);
+        window_listbox.insert(new_item, -1);
+        // prevent new_item's parent(ListBoxRow) to be focusable
+        new_item.get_parent().set_can_focus(false);
+        new_item.get_parent().show_all();
+
+        return new_item;
+    }
+
+    public void remove_window(ApplicationWindow appwin)
+    {
+        var list_row = window_row_map [appwin];
+
+        // Handles the removal, with animation, of WindowListItem from the ListBox
+        list_row.gracefully_die(window_listbox);
+        window_row_map.remove(appwin);
+        // hold item to prevent its destruction and the possible resize of the
```

Comment:
It is a habit thing. The multi-line comments make it impossible to comment out sections of the code, which I do from time to time for debugging.
I can change that to the multi-line ones if you prefer.